### PR TITLE
Warn on using literals in tight packing

### LINKED ISSUE #2543: Show warning for non-typed constant in tight packing
Part of #982.

```
contract A {
  function f() {
    keccak256(1);
  }
}
```

Show warning that `1` is considered `uint8` and use an explicit type if otherwise desired (similar to the warning of `var`).

### PR DESCRIPTION
**Pull request by axic**
Fixes #2543.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1446,6 +1446,32 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		_functionCall.annotation().type = make_shared<TupleType>(functionType->returnParameterTypes());
 
 	TypePointers parameterTypes = functionType->parameterTypes();
+
+	if (
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Perhaps better to use `!_functionType.padArguments()`

Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: Updated.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -1446,6 +1446,28 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		_functionCall.annotation().type = make_shared<TupleType>(functionType->returnParameterTypes());
 
 	TypePointers parameterTypes = functionType->parameterTypes();
+
+	if (!_functionType.padArguments())
+	{
+		for (size_t i = 0; i < arguments.size(); ++i)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Could use range-based loop, but is also fine this way.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -1446,6 +1446,28 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
...
+		{
+			auto const& argType = type(*arguments[i]);
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: this variable is not really required, you can just move that into the argument of the dynamic_cast.

**Review by chriseth — APPROVED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: This file makes heavy use of this, probably should just make one giant cleanup PR changing all?

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: `argType` is also used in the message.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Sure, but you can as well call the function on `literal` instead of on `argType`.

**Pull request merged**